geomesa-hbase/geomesa-hbase-datastore/src/test/scala/org/locationtech/geomesa/hbase/data/HBaseTTLTest.scala

```diff
@@ -0,0 +1,135 @@
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.hbase.data
+
+import java.text.SimpleDateFormat
+import java.util.Date
+
+import com.typesafe.scalalogging.LazyLogging
+import org.geotools.data._
+import org.geotools.data.collection.ListFeatureCollection
+import org.geotools.data.simple.SimpleFeatureStore
+import org.geotools.filter.text.ecql.ECQL
+import org.geotools.util.factory.Hints
+import org.junit.runner.RunWith
+import org.locationtech.geomesa.features.ScalaSimpleFeature
+import org.locationtech.geomesa.hbase.data.HBaseDataStoreParams._
+import org.locationtech.geomesa.utils.collection.SelfClosingIterator
+import org.locationtech.geomesa.utils.conf.FeatureExpiration
+import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypes
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+import org.specs2.mutable.Specification
+import org.specs2.runner.JUnitRunner
+
+import scala.collection.JavaConversions._
+import scala.collection.JavaConverters._
+import scala.concurrent.duration.Duration
+
+@RunWith(classOf[JUnitRunner])
+class HBaseTTLTest extends Specification with LazyLogging {
+
+  sequential
+
+  val typeName = "test_data"
+  val params = Map(
+    ConnectionParam.getName -> MiniCluster.connection,
+    HBaseCatalogParam.getName -> getClass.getSimpleName)
+
+  val expiration = "4 seconds"
+  val sft: SimpleFeatureType = SimpleFeatureTypes.createType(typeName, f"name:String:index=true,dtg:Date;geomesa.feature.expiry=$expiration")
+  // NOTE: expiration date is set simply by adding expiration time to when the feature is created
+  val ttl = Duration(expiration).toMillis
+
+  val dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ")
+  "HBase TTL" should {
+    "remove X features based on TTL" >> {
+      val numFeatures = 3
+      val ds = DataStoreFinder.getDataStore(params).asInstanceOf[HBaseDataStore]
+      ds must not(beNull)
+      ds.createSchema(sft)
+
+      try {
+        val fs = ds.getFeatureSource(typeName).asInstanceOf[SimpleFeatureStore]
+
+        val startTime = new Date(System.currentTimeMillis())
+
+        val toAdd = (0 until numFeatures).map { i =>
+          val sf = new ScalaSimpleFeature(sft, i.toString)
+          sf.getUserData.put(Hints.USE_PROVIDED_FID, java.lang.Boolean.TRUE)
+          sf.setAttribute(0, s"name$i")
+
+          val featureTime = new Date(startTime.getTime + (ttl / 2) * i)
+          sf.setAttribute(1, dateFormat.format(featureTime))
+          sf
+        }
+
+        val ids = fs.addFeatures(new ListFeatureCollection(sft, toAdd))
+        ids.asScala.map(_.getID) must containTheSameElementsAs((0 until numFeatures).map(_.toString))
+
+        getElements(ds, typeName).size mustEqual numFeatures
+
+        Thread.sleep(ttl) // wait until one feature times out
```

Comment:
hmm, i wish there was a better way to test these without relying on timing, but nothing is coming to mind...

Reply:
My terrible idea was using AOP to intercept the system get time call with aspectj or something.

Reply:
there is an 'eventually' call in specs that will re-try something a configurable number of times, that may make this more robust in case the hbase ttl doesn't always get processed right away - but maybe it always would? i'm not sure on the internals of how hbase processes ttls. and you'd still need to sleep to let it expire.
for reference: https://github.com/locationtech/geomesa/blob/main/geomesa-kafka/geomesa-kafka-datastore/src/test/scala/org/locationtech/geomesa/kafka/data/KafkaDataStoreTest.scala#L170

Reply:
some frameworks also let you inject a clock that you can control, but i doubt that would be feasible using the hbase mini cluster